tests/main/views/test_drafts.py

```diff
@@ -56,6 +62,455 @@ def service_count(self):
+        res1 = self.client.put(
+            '/draft-services/copy-from/{}'.format(self.service_id),
+            data=json.dumps(self.updater_json),
+            content_type='application/json')
+
+        assert res1.status_code == 201
+
+        res2 = self.client.put(
+            '/draft-services/copy-from/{}'.format(self.service_id),
+            data=json.dumps({
+                **self.updater_json,
+                **self.basic_questions_to_copy,
+                'targetFramework': 'g-cloud-7'
+            }),
+            content_type='application/json')
+
+        assert res2.status_code == 201
+
+    def test_only_specified_questions_are_copied(self):
+        res = self.client.put(
+            '/draft-services/copy-from/{}'.format(self.service_id),
+            data=json.dumps({
+                **self.updater_json,
+                'targetFramework': 'g-cloud-7',
+                'questionsToCopy': ['serviceBenefits', 'termsAndConditionsDocumentURL']
+            }),
+            content_type='application/json')
+        assert res.status_code == 201
+
+        draft = DraftService.query.filter(
+            DraftService.framework_id == 4,
+        ).first()
+
+        assert {'serviceBenefits', 'termsAndConditionsDocumentURL'} == set(draft.data.keys())
+
+    def test_source_service_is_marked_as_copied_after_copy_to_new_framework(self):
+        pre_copy_source_service = Service.query.first()
+        assert pre_copy_source_service.copied_to_following_framework is False
+
+        res = self.client.put(
+            '/draft-services/copy-from/{}'.format(self.service_id),
+            data=json.dumps({
+                **self.updater_json,
+                **self.basic_questions_to_copy,
+                'targetFramework': 'g-cloud-7'
+            }),
+            content_type='application/json')
+        assert res.status_code == 201
+
+        post_copy_source_service = Service.query.first()
+        assert post_copy_source_service.copied_to_following_framework is True
+
+    @pytest.mark.parametrize('framework_status', Framework.STATUSES)
+    def test_services_can_not_be_copied_to_a_framework_that_is_not_open(self, framework_status):
+        if framework_status == 'open':
+            return
+
+        self.set_framework_status('g-cloud-7', framework_status)
+
+        res = self.client.put(
+            '/draft-services/copy-from/{}'.format(self.service_id),
+            data=json.dumps({
+                **self.updater_json,
+                **self.basic_questions_to_copy,
+                'targetFramework': 'g-cloud-7'
+            }),
+            content_type='application/json')
+
+        assert res.status_code == 400
+        assert 'Target framework is not open' in res.get_data(as_text=True)
+
+
+class TestCopyPublishedFromFramework(DraftsHelpersMixin):
+    def post_to_copy_published_from_framework(
+        self, source_framework_slug="g-cloud-6", questions_to_copy='default', framework_slug='g-cloud-7'
```

Comment:
The default value here for `questions_to_copy` could be slightly confusing as we're providing a string, but if you want to override the value you'll need to provide a list. Without type hints here, it's not immediately obvious how to use this parameter correctly. A docstring would help, but I think maybe we could rename the variable `override_questions_to_copy` and have the default as None. Having the available types as `(None, list)` is a bit more standard than `(str, list)`, IMO. Especially because strings and lists are both iterables, but you generally don't want to iterate over a string - so it's better to have that be a None which will break when you try to iterate over it.

Reply:
This reminds me that I still do want to add type hints to our codebase one day...

Reply:
Also, `framework_slug` should probably be called `target_framework_slug` for the avoidance of any doubt (given we have a `source_framework_slug`).

Reply:
Yeah, it felt a bit funky when I was writing it. I like your suggestions. Have updated, with a docstring to be really clear.

Reply:
OK, I remember why I did it this way. Having the default for the kwarg as `None`, which then means that the `basic_questions_json` is used, makes it difficult to test the case when I actually want to test the case when `questions_to_copy` isn't supplied at all.

I've added a special case for when `override_questions_to_copy` is set to 'No questions' so that the `questionsToCopy` key can be skipped from the request data completely. Have explained in the docstring. This _might_  be a terrible idea. Probably is. Interested to know what you think.